File: codewars/6kyu/which-are-in/dduck-which-are-in.py
# which-are-in
# https://www.codewars.com/kata/550554fd08b86f84fe000a58/

from unittest import TestCase


def in_array(array1, array2):
    return sorted(
        filter(
            lambda x: max(
                list(
                    map(lambda y: y.find(x), array2)
                ) + [-1]
            ) > -1,
            list(set(array1))
        )
    )


# it's imitation of JavaScript best practice

a1 = ["live", "arp", "strong"]
a2 = ["lively", "alive", "harp", "sharp", "armstrong"]
r = ['arp', 'live', 'strong']
TestCase().assertEqual(in_array(a1, a2), r)

a1 = ['arp', 'mice', 'bull']
a2 = ['lively', 'alive', 'harp', 'sharp', 'armstrong']
r = ['arp']
TestCase().assertEqual(in_array(a1, a2), r)

a1 = ['duplicates', 'duplicates']
a2 = ['duplicates', 'duplicates']
r = ['duplicates']
TestCase().assertEqual(in_array(a1, a2), r)

File: codewars/6kyu/which-are-in/dduck-which-are-in.js
// which-are-in
// https://www.codewars.com/kata/550554fd08b86f84fe000a58/

const inArray = (array1, array2) => {
  return array1.map((array1Element) => {
    return array2.map((array2Element) => {
      return array2Element.search(array1Element);
    }).filter((v) => v >= 0).length ? array1Element : null;
  }).filter((v) => v)
    .sort();
};
// I really made trash code today.
// shame on me.

// best practice in codewars
// const inArray = (arr1, arr2) => {
//   return arr1.filter((arr1v) => {
//     return arr2.some((arr2v) => {
//       return arr2v.search(arr1v) > -1;
//     });
//   }).sort();
// };

const a2 = ["lively", "alive", "harp", "sharp", "armstrong"];

let a1 = ["xyz", "live", "strong"];
console.log(inArray(a1, a2), ["live", "strong"]);

a1 = ["live", "strong", "arp"];
console.log(inArray(a1, a2), ["arp", "live", "strong"]);

a1 = ["tarp", "mice", "bull"];
console.log(inArray(a1, a2), []);
